File: src/lib/logSanitizer.ts
/**
 * logSanitizer.ts
 * Redacts sensitive values (API keys, tokens, passwords, JWTs) from log strings.
 * Port of log_sanitizer.py for the Next.js/Node.js runtime.
 */

const REPLACEMENT = '[REDACTED]';

interface SanitizeRule {
    name: string;
    pattern: RegExp;
    replace: (match: RegExpMatchArray) => string;
}

const DEFAULT_RULES: SanitizeRule[] = [
    // Authorization headers: keep scheme, redact credential
    {
        name: 'auth_header',
        pattern: /(authorization\s*:\s*)(bearer|basic|token)\s+([^\s,;]+)/gi,
        replace: (m) => `${m[1]}${m[2]} ${REPLACEMENT}`,
    },
    // key=value / key: value — keep key, redact value
    {
        name: 'api_key_kv',
        pattern: /\b(api[_-]?key|token|secret|password|passwd|pwd)\b\s*[:=]\s*(["']?)([^"'\s]{8,})\2/gi,
        replace: (m) => `${m[1]}: ${REPLACEMENT}`,
    },
    // URL query params: keep param name, redact value
    {
        name: 'api_key_qs',
        pattern: /([?&](api[_-]?key|token|access[_-]?token|secret)=)([^&\s]+)/gi,
        replace: (m) => `${m[1]}${REPLACEMENT}`,
    },
    // JWT-like strings: fully redact
    {
        name: 'jwt',
        pattern: /\beyJ[a-zA-Z0-9_-]{10,}\.[a-zA-Z0-9_-]{10,}\.[a-zA-Z0-9_-]{10,}\b/g,
        replace: () => REPLACEMENT,
    },
];

/**
 * Sanitizes a string by redacting sensitive patterns.
 */
export function sanitizeText(text: string, rules: SanitizeRule[] = DEFAULT_RULES): string {
    if (!text) return text;

    let out = text;
    for (const rule of rules) {
        // Reset lastIndex for global regexes
        rule.pattern.lastIndex = 0;
        out = out.replace(rule.pattern, (...args) => {
            // args: full match, ...groups, offset, string
            const groups = args.slice(0, -2);
            return rule.replace(groups as RegExpMatchArray);
        });
    }
    return out;
}

/**
 * Wraps console methods to sanitize output.
 * Call once at app startup (e.g., in next.config.ts instrumentation or middleware).
 */
export function installSanitizingLogger(): void {
    const methods = ['log', 'warn', 'error', 'info', 'debug'] as const;

    for (const method of methods) {
        const original = console[method].bind(console);
        (console[method] as (...args: unknown[]) => void) = (...args: unknown[]) => {
            const sanitized = args.map((arg) =>
                typeof arg === 'string' ? sanitizeText(arg) : arg
            );
            original(...sanitized);
        };
    }
}

File: security/log_sanitizer.py
# log_sanitizer.py
import logging
import re
from typing import Pattern, List, Tuple

REPLACEMENT = "[REDACTED]"

DEFAULT_RULES: List[Tuple[str, Pattern[str]]] = [
    # Authorization headers (Bearer, Basic, Token)
    ("auth_header", re.compile(r"(?i)(authorization\s*:\s*)(bearer|basic|token)\s+([^\s,;]+)")),
    # Common key/value patterns
    ("api_key_kv", re.compile(r"(?i)\b(api[_-]?key|token|secret|password|passwd|pwd)\b\s*[:=]\s*([\"']?)([^\"'\s]{8,})(\2)")),
    # URL query params
    ("api_key_qs", re.compile(r"(?i)([?&](api[_-]?key|token|access[_-]?token|secret)=)([^&\s]+)")),
    # JWT-like strings (very common in logs)
    ("jwt", re.compile(r"\beyJ[a-zA-Z0-9_-]{10,}\.[a-zA-Z0-9_-]{10,}\.[a-zA-Z0-9_-]{10,}\b")),
]

def sanitize_text(text: str, rules=DEFAULT_RULES) -> str:
    if not text:
        return text

    out = text

    # Authorization header: keep scheme, redact credential
    out = DEFAULT_RULES[0][1].sub(lambda m: f"{m.group(1)}{m.group(2)} {REPLACEMENT}", out)

    # key=value (keep key, redact value)
    out = DEFAULT_RULES[1][1].sub(lambda m: f"{m.group(1)}: {REPLACEMENT}", out)

    # query string: keep param name, redact value
    out = DEFAULT_RULES[2][1].sub(lambda m: f"{m.group(1)}{REPLACEMENT}", out)

    # jwt: fully redact
    out = DEFAULT_RULES[3][1].sub(REPLACEMENT, out)

    return out

class SanitizingFormatter(logging.Formatter):
    def format(self, record: logging.LogRecord) -> str:
        msg = super().format(record)
        return sanitize_text(msg)

def install_sanitizing_formatter(logger_name: str = "", level=logging.INFO) -> None:
    """
    Installs sanitizer on root logger by default.
    """
    logger = logging.getLogger(logger_name)
    logger.setLevel(level)

    for handler in logger.handlers:
        fmt = handler.formatter or logging.Formatter("%(asctime)s %(levelname)s %(name)s %(message)s")
        handler.setFormatter(SanitizingFormatter(fmt._fmt))

# Example usage:
# import logging
# from log_sanitizer import install_sanitizing_formatter
# logging.basicConfig(level=logging.INFO)
# install_sanitizing_formatter()
# logging.info("Authorization: Bearer abcdef123456")
